plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/codeInsight/unwrap/KotlinFunctionParameterUnwrapper.kt

```diff
@@ -3,25 +3,176 @@
+     *
+     */
+    private fun findAdjacentValueArgumentInsideParen(element: PsiElement): KtValueArgument? {
+        if (element.parent !is KtValueArgumentList) {
+            return null
+        }
+
+        return when {
+            element.elementType == KtTokens.RPAR -> {
+                val valueArgumentList = element.parent as? KtValueArgumentList ?: return null
+                if (valueArgumentList.parent !is KtCallExpression) return null
+                return valueArgumentList.arguments.lastOrNull()
+            }
+            element.isComma -> {
+                // caret before comma, choose previous argument
+                val previous = element.getPrevSiblingIgnoringWhitespaceAndComments()
+                isCallArgument(previous).ifTrue { previous as KtValueArgument }
+            }
+            element is PsiWhiteSpace -> {
+                // caret before blank, argument could be towards the left or the right
+                val previous = element.getPrevSiblingIgnoringWhitespaceAndComments()
+                if (isCallArgument(previous)) return previous as KtValueArgument
+
+                val next = element.getNextSiblingIgnoringWhitespaceAndComments()
+                if (isCallArgument(next)) return next as KtValueArgument
+
+                null
+            }
+            else -> null
+        }
+    }
+
+    private fun isCallArgument(element: PsiElement?): Boolean {
+        if (element is KtLambdaArgument) return false
+        if (element !is KtValueArgument) return false
+        val argumentList = element.parent as KtValueArgumentList
+        if (argumentList.parent !is KtCallExpression) return false
+
         return true
     }
 
     override fun doUnwrap(element: PsiElement?, context: Context?) {
-        val function = element as? KtCallExpression ?: return
-        val argument = element.valueArguments.firstOrNull()?.getArgumentExpression() ?: return
-        context?.extractFromExpression(argument, function)
-        context?.delete(function)
+        val valueArgument = argumentToUnwrap(element) ?: return
+        val deletionTarget = deletionTarget(valueArgument) ?: return
+        val argument = valueArgument.getArgumentExpression() ?: return
+        context?.extractFromExpression(argument, deletionTarget)
+        context?.delete(deletionTarget)
     }
 
+    private fun deletionTarget(valueArgument: KtValueArgument): KtElement? {
+        var function: KtElement = valueArgument.getStrictParentOfType<KtCallExpression>() ?: return null
+        val parent = function.parent
+        if (parent is KtQualifiedExpression) {
+            function = parent
+        }
+        return function
+    }
+
+    override fun collectAffectedElements(e: PsiElement, toExtract: MutableList<PsiElement>): PsiElement {
+        val valueArgument = argumentToUnwrap(e) ?: return e
+        val deletionTarget = deletionTarget(valueArgument) ?: e
+
+        toExtract.add(valueArgument)
+
+        return deletionTarget
```

Comment:
```suggestion
        super.collectAffectedElements(e, toExtract)
        return argumentToUnwrap(e)?.let { deletionTarget(it) } ?: e
```

Reply:
For same reasons as `collectAffectedElements` in [KotlinLambdaUnwrapper](https://github.com/JetBrains/intellij-community/pull/1996/files#diff-a249d4cd35f4ed08fe5c4342e64d9d275f142bb0c01e5f9e918f7f6a1fb1cc18)

Reply:
Done